mlir/lib/IR/HugrTypeInterfaces.cpp

```diff
@@ -0,0 +1,129 @@
+          ExternalIndexHugrTypeInterfaceModel, mlir::IndexType> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return TypeConstraint::Copyable;
+  }
+};
+
+struct ExternalTupleHugrTypeInterfaceModel
+    : public HugrTypeInterface::ExternalModel<
+          ExternalTupleHugrTypeInterfaceModel, mlir::TupleType> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return TypeConstraint::Copyable;
+  }
+};
+
+struct ExternalIntegerHugrTypeInterfaceModel
+    : public HugrTypeInterface::ExternalModel<
+          ExternalIntegerHugrTypeInterfaceModel, mlir::IntegerType> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return TypeConstraint::Copyable;
+  }
+};
+
+struct ExternalFloat32HugrTypeInterfaceModel
+    : public HugrTypeInterface::ExternalModel<
+          ExternalFloat32HugrTypeInterfaceModel, mlir::Float32Type> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return TypeConstraint::Copyable;
+  }
+};
+
+struct ExternalFloat64HugrTypeInterfaceModel
+    : public HugrTypeInterface::ExternalModel<
+          ExternalFloat64HugrTypeInterfaceModel, mlir::Float64Type> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return TypeConstraint::Copyable;
+  }
+};
+
+struct SumHugrTypeInterfaceModel : public HugrTypeInterface::ExternalModel<
+                                       SumHugrTypeInterfaceModel, SumType> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return TypeConstraint::Copyable;
+  }
+};
+
+struct OpaqueHugrTypeInterfaceModel
+    : public HugrTypeInterface::ExternalModel<
+          OpaqueHugrTypeInterfaceModel, OpaqueType> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return ExtensionSetAttr::get(t.getContext());
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return llvm::cast<AliasRefType>(t).getConstraint();
+  }
+};
+
+struct AliasRefHugrTypeInterfaceModel
+    : public HugrTypeInterface::ExternalModel<
+          AliasRefHugrTypeInterfaceModel, AliasRefType> {
+  ExtensionSetAttr getExtensions(mlir::Type t) const {
+    return llvm::cast<AliasRefType>(t).getExtensions();
+  }
+  TypeConstraint getConstraint(mlir::Type t) const {
+    return llvm::cast<AliasRefType>(t).getConstraint();
+  }
+};
+
+struct ExtendedHugrTypeInterfaceModel
```

Comment:
I don't think ExtendedType is a good name for this, but I can't think of a better one OTTOMH